edgelet/aziot-edged/src/image_gc.rs

```diff
@@ -0,0 +1,265 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+use std::{collections::HashSet, time::Duration};
+
+use crate::error::Error as EdgedError;
+use chrono::Timelike;
+use edgelet_core::{ModuleRegistry, ModuleRuntime};
+use edgelet_docker::ImagePruneData;
+use edgelet_settings::{base::image::ImagePruneSettings, RuntimeSettings};
+
+const TOTAL_MINS_IN_DAY: u32 = 1440;
+
+pub(crate) async fn image_garbage_collect(
+    settings: edgelet_settings::Settings,
+    runtime: &edgelet_docker::DockerModuleRuntime<http_common::Connector>,
+    image_use_data: ImagePruneData,
+) -> Result<(), EdgedError> {
+    log::info!("Starting image auto-pruning task...");
+
+    let edge_agent_bootstrap: String = settings.agent().config().image().to_string();
+
+    // bootstrap edge agent image should never be deleted
+    let bootstrap_image_id: String = match ModuleRuntime::list_images(runtime).await {
+        Ok(image_name_to_id) => image_name_to_id
+            .get(&edge_agent_bootstrap)
```

Comment:
Hmmm.. this bears thinking about.

Reasoning through, this is what I believe will happen:

- Right now, we keep the image_ids associated with 'docker ps -a' in a Set.
- To this set, we add the CURRENT bootstrap image_id (which we get by parsing the current config.toml).
- What happens behind the scenes is that all of the images in the set are recorded in our persistence file with the timestamp System::now()/current_time.
- So we know that all these images -- BOOTSTRAP image included -- exist, and that we have to manage them.

Presumably, for newer versions, the config.toml would be updated (if it's not, we don't face the issue of bootstrap images piling up). So, after edged restart, the OLD bootstrap image would no longer be parsed from config.toml. Instead, it will be the new one. Which means that the timestamp associated with the old bootstrap image will not be updated in the persistence file. Eventually, that image will be older than the "min age" and will be cleaned up.

That said - I think my variable names make it hard to see this right away, and I'll change some to make this more straightforward to understand.